src/openms/source/CHEMISTRY/AASequence.cpp

```diff
@@ -908,44 +908,55 @@ namespace OpenMS
       ++next_aa;
       if (*next_aa == '.') ++next_aa;
 
-      aas.n_term_mod_ = &(mod_db->getModification(mod, String(*next_aa), ResidueModification::N_TERM));
+      const ResidueModification* m = mod_db->getModification(mod, String(*next_aa), ResidueModification::N_TERM);
+      if (m == nullptr)
+      {
+        throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, str,
+          "Cannot convert string to peptide modification. No N-terminal modification matches to term specificty and origin.");
+      }
+      aas.n_term_mod_ = m;
       return mod_end;
     }
 
     const String& res = aas.peptide_.back()->getOneLetterCode();
     if (specificity == ResidueModification::C_TERM) 
     {
-      aas.c_term_mod_ = &(mod_db->getModification(mod, res, ResidueModification::C_TERM));
+      const ResidueModification* m = mod_db->getModification(mod, res, ResidueModification::C_TERM);
+      if (m == nullptr)
+      {
+        throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, str,
+          "Cannot convert string to peptide modification. No C-terminal modification matches to term specificty and origin.");
+      }
+      aas.c_term_mod_ = m;
       return mod_end;
     }
 
-    try
+    const Residue* internal = ResidueDB::getInstance()->getModifiedResidue(aas.peptide_.back(), mod);
+    if (internal != nullptr)
     {
-      aas.peptide_.back() = ResidueDB::getInstance()->getModifiedResidue(aas.peptide_.back(), mod);
+      aas.peptide_.back() = internal;
     }
-    catch (Exception::InvalidValue) // no such mod for this residue
+    else // no internal mod for this residue
     {
       // TODO: get rid of this code path, its deprecated and is only a hack for
       // C-terminal modifications that don't use the dot notation
       if (std::distance(mod_end, str.end()) == 1) // potentially a C-terminal mod without explicitly declaring it using dot notation?
       {
         // old ambiguous notation: Modification might be at last amino acid or at C-terminus
-        try
+        const ResidueModification* term_mod = mod_db->getModification(mod, res, ResidueModification::C_TERM);
+        aas.c_term_mod_ = term_mod;
```

Comment:
@timosachsenberg this is not quite what was there before (before c_term_mod was only set when no exception occured). Not sure whether that makes a difference, but maybe move this line to `if (term_mod != nullptr)`

Reply:
it sets c_term_mod_ to nullptr in that case which should be the same as before. but yes it might be easier to read if I put it in an extra if block